### ISSUE TITLE
Document the exact meanings of crate-types and library lookups

### ISSUE THREAD

**Issue by metux** [rust-lang/rust#43191]
The documentation on the various crate-types is very poor, especially things we need for distro packaging. 

For example, there'e different types of shared objects, probably one w/ metadata and one w/o.
In binpkg-distros we need runtime (actual binary code), debug-info and build-time stuff separated
(so, they can installed/removed separately, but dedicated packages).

Other vital issue is ABI compatibility between shared object versions - we need at least soname's,
symver's would be even better.

**Comment by steveklabnik** [comment]
> Other vital issue is ABI compatibility between shared object versions

We don't make any ABI guarantees at all, see https://github.com/rust-lang/rfcs/issues/600

**Comment by jmesmon** [comment]
Out of curiosity, what distribution are you writing/modifying packaging for?

It may help to look at the approach other distros, like [fedora](https://src.fedoraproject.org/cgit/rpms/rust.git/tree/rust.spec), [debian](https://anonscm.debian.org/cgit/pkg-rust/rust.git/tree/), and [arch](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/rust) have taken in packaging (IIRC, there are also the source based distros gentoo & openembedded to look at, but they may not have the same needs).

In general, binaries written in rust being dynamically linked against libraries written in rust is not very common (afaik, only components of this repository do that). The result is that the need for considering the rustc libraries as "runtime" for other rust-based programs/libraries is very rare.

The dynamic libraries shipped with rustc are, afaik, not intended to provide any stable interface and are primarily used as a internal mechanism to share code between `rustc`, `rustdoc`, and other components.

As for splitting debug-info, I know that in openembedded we have the packaging system split it to a separate set of files completely separate from rust's build process. I don't know how other distros handle things.

**Comment by cuviper** [comment]
For Fedora's part, we only ship the shared libraries for use by `rustc` and `rustdoc`, and for transitively-built compiler plugins (`proc_macro` derives).  We explicitly filter them from rpm auto-Provides for lack of ABI.

**Comment by steveklabnik** [comment]
Triage: p-medium. This is likely to go in the Cargo Book.

**Comment by DevQps** [comment]
@steveklabnik I am not sure but [here](https://doc.rust-lang.org/reference/linkage.html) is already some documentation on linkage and the various type of crate-types. I am not sure if this is sufficient enough to close this issue? It seems quite comprehensive.

**Comment by steveklabnik** [comment]
Yes, let's close this issue in favor of that; @metux if you have specific questions that still aren't answered, please open up issues on the reference for each question. Thanks!

**Comment by metux** [comment]
Hello folks,

Sorry, I've missed this issue for long time (too busy w/ other things and had to put rust aside).
With distros we have several problems to solve:

a) We need full control of the whole build process, eg. which toolchain versions are used, exact sources, ...
b) We need to be able to patch all components anytime, w/o any coorperation from upstreams - we have to do hotfixes.
c) We need to be able to rebuild everything from source at any time.
d) We need to have our own copies of the source trees (can't rely on downloding from somewhere else - actually, cant download anything except by the distro's package manager itself)
e) We need to make sure that everything still fits together even after rebuilds and upgrades
f) We need debug info, build-time stuff, etc, in separate files, so it can be deployed via separate packages only when needed
g) We need to make sure that fixes in some package really get active in all other depending packages (shared libraries give that for free)
h) We need to take care of the resource consumption of a whole system (many applications installed and running in parallel)

Meanwhile I've learned that shared libraries aren't a good idea w/ rust. But this opens up another nasty problem: we have to make sure that dozens of depending packages need to be rebuilt when 
a bug in a library has been fixed (and find out what needs a rebuild in the first place).

And there's also the open problem of separating out debug info, build-time only stuff, ...

--mtx

**Comment by DevQps** [comment]
@metux Glad to hear your response! As a comment on some of your points (in a bit different order):

c) We need to be able to rebuild everything from source at any time.
- Each crate from crates.io currently gets downloaded as an archived (zip) file with the source code inside. In that sense you can always rebuild things from source.

b) We need to be able to patch all components anytime, w/o any coorperation from upstreams - we have to do hotfixes.
- If you need to do hotfixes you can always fork the original repository, create a hotfix, and then reference to it by `somecrate = {git = "somepathtoabranch"}. See [Dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html) in the Cargo book for more.

a) We need full control of the whole build process, eg. which toolchain versions are used, exact sources.
Since a new application is built from archived sources from crates.io or from git repositories you always build from source (unless you have some files cached), with the toolchain of your choice.

If you still find some parts lacking, then please open some issues for each separate part! Then we'll try to figure out whether or not we can improve some of the parts you mentioned. Good luck with everything!

**Comment by metux** [comment]
Hi,

> c) We need to be able to rebuild everything from source at any time.
> 
> * Each crate from crates.io currently gets downloaded as an archived (zip) file with the source code inside. In that sense you can always rebuild things from source.

Unfotunately, that's not enough.

Instead I need to prevent it from attempting to download anything, but just take everything from already prepared sysroot image (where the distro's build system deploys dependencies, before starting the actual build. For that, these deps already have to be packaged.

> b) We need to be able to patch all components anytime, w/o any coorperation from upstreams - we have to do hotfixes.
> 
> * If you need to do hotfixes you can always fork the original repository, create a hotfix, and then reference to it by `somecrate = {git = "somepathtoabranch"}. See [Dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html) in the Cargo book for more.

In that context, we can't pull from any external git repo, nor download anything (build jails usually don't even have free network access, for good reasons). Anything needed by the build process (except for the actual source tree) has to be deployed exclusively via package manager (by the deps declared in the control file). 

Having to patch all consumer packages for a specific dependency version is very impractical (would need to handle NxN relationships, thus NxN people have to corrdinate in short time) - nothing we could practically do in distros w/ ~100k packages.

Also note that certain metadata, eg. package name/version, dependencies, etc, need to be extracted *outside* the build jails, which tend to run on entirely different distro versions. Therefore assuming specific rust version (or even rust at all) in the host (instead of specific jails) is even more impractical and would rule out a larga porting of currently used build host in the field, ergo: destroys reproducability.
 
> a) We need full control of the whole build process, eg. which toolchain versions are used, exact sources.
> 
> * Since a new application is built from archived sources from crates.io or from git repositories you always build from source (unless you have some files cached), with the toolchain of your choice.

That's exactly what we cannot have - can't rely on any external locations whatsoever - everything needs to be in either source or (already built) binary packages.

In summary: the whole idea of pulling everything from some external repo (same for pypy, mvn, etc) is really bad for distros :(

**Comment by DevQps** [comment]
Ahh I am sorry then! I don't know too much about distro's so I'll leave that up to other people :)

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by steveklabnik** [comment]
@metux I know you said you don't have time for Rust anymore, so I'm not sure if this is an active area of research for you or not, but Debian, Fedora, and others have solved a lot of these problems already, with the same constraints. I'd encourage you or anyone else interested in this topic to reach out to them with answers to these kinds of questions.
